tests/test_optimize.py

```diff
@@ -462,6 +468,113 @@ def test_adam_optimizer_multivar(self, bunch, tol):
+                x_vec = x_vals[jdx:jdx + 2]
+
+                x_onestep = bunch.rotosolve_opt.step(f, x_vec)
+                x_onestep_target = self.rotosolve_step(f, x_vec)
+
+                assert x_onestep == pytest.approx(x_onestep_target, abs=tol)
+
+                x_twosteps = bunch.rotosolve_opt.step(f, x_onestep)
+                x_twosteps_target = self.rotosolve_step(f, x_onestep_target)
+
+                assert x_twosteps == pytest.approx(x_twosteps_target, abs=tol)
+
+    @pytest.mark.parametrize('x_start', [0.5, -1.14, [0.3, 0.25], [-0.6, 0.45], [1.3, -0.9]])
+    @pytest.mark.parametrize('generators', [list(tup) for tup in it.product([qml.RX, qml.RY, qml.RZ], repeat=2)])
+    def test_rotoselect_optimizer(self, x_start, generators, bunch, tol):
+        """Tests that rotoselect optimizer takes one and two steps correctly for the VQE circuit
+        defined in `this rotoselect tutorial <https://pennylane.ai/qml/demos/tutorial_rotoselect.html>`_."""
+
+        # the optimal generators for the 2-qubit VQE circuit
+        # H = 0.5 * Y_2 + 0.8 * Z_1 - 0.2 * X_1
+        optimal_generators = [qml.RY, qml.RX]
+        possible_generators = [qml.RX, qml.RY, qml.RZ]
+        bunch.rotoselect_opt.possible_generators = possible_generators
+
+        dev = qml.device("default.qubit", analytic=True, wires=2)
+
+        def ansatz_rsel(params, generators):
+            generators[0](params[0], wires=0)
+            generators[1](params[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+
+        @qml.qnode(dev)
+        def circuit_rsel(params, generators=None):  # generators will be passed as a keyword arg
+            ansatz_rsel(params, generators)
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))
+
+        @qml.qnode(dev)
+        def circuit_rsel2(params, generators=None):  # generators will be passed as a keyword arg
+            ansatz_rsel(params, generators)
+            return qml.expval(qml.PauliX(0))
+
+        def cost_fn(params, generators):
+            # test single parameter inputs
+            if np.ndim(params) == 1:
+                params = [params[0], params[0]]
+            Z_1, Y_2 = circuit_rsel(params, generators=generators)
+            X_1 = circuit_rsel2(params, generators=generators)
+            return 0.5 * Y_2 + 0.8 * Z_1 - 0.2 * X_1
+
+        x_onestep, generators = bunch.rotoselect_opt.step(cost_fn, x_start, generators)
+        f_best_gen = lambda x_start: cost_fn(x_start, generators=optimal_generators)
+        x_onestep_target = self.rotosolve_step(f_best_gen, x_start)
+
+        assert x_onestep == pytest.approx(x_onestep_target, abs=tol)
+
+        x_twosteps, generators = bunch.rotoselect_opt.step(cost_fn, x_onestep, generators)
+        x_twosteps_target = self.rotosolve_step(f_best_gen, x_onestep_target)
+
+        assert x_twosteps == pytest.approx(x_twosteps_target, abs=tol)
+
+    @staticmethod
+    def rotosolve_step(f, x):
```

Comment:
It might be worth moving this to the top of the relevant tests

Reply:
Is it better to have it above the relevant tests (not at the top of the file) rather than below them? Or do you mean that I should put a copy of ``rotosolve_step`` inside every test that uses it?

Reply:
Just to have it above the relevant tests :)

Reply:
I think best to move this static method to above the first test `test_rotosolve_optimizer_univar`, so that developers are already familiar with this helper method :)